### ISSUE TITLE
Skip file on error in zip_source_function callback

### ISSUE THREAD

**Issue by chulup** [nih-at/libzip#145]
I am using `zip_source_function` as data source for `zip_file_add` and sometimes on ZIP_SOURCE_OPEN file (data source) could not be read which means as lawful developer I call `zip_error_set` and return -1. Of course, libzip immediately drops everything and returns from `zip_close` with an error.

How can I tell libzip to skip that file or at least leave it empty but not cancel everything?

On somewhat different topic: I use `zip_source_function` because sometimes I need to stop `zip_close` and cancel operation (just as another issue #135 ), which I can do by returning -1 from source callback. Would be nice to be able to use `zip_progress_callback` for that or have any other means of interrupting the process.

**Comment by 0-wiz-0** [comment]
This functionality does not exist, and I'm not sure we'll add it. (Currently I suspect rather not.)
One easy workaround is to supply 0 bytes of data and call it a day (i.e. return success).

As for your second request – you already found https://github.com/nih-at/libzip/issues/135 and we've actually already started work on it, but will need some more time to debug and finish it.

**Comment by 0-wiz-0** [comment]
We've implemented cancel support.

An alternative suggestion for your problem:
return -1 as of now and save some state information;
when `zip_close()` returns an error, call `zip_delete()` to remove the offending entry and try again.

**Comment by EZ-Phantom** [comment]
> We've implemented cancel support.

And can you tell when it will be in release verison? Will be very kind from your side, thanks!

**Comment by chulup** [comment]
> 
> 
> > We've implemented cancel support.
> 
> And can you tell when it will be in release verison? Will be very kind from your side, thanks!

@EZ-Phantom 
You can use my copy of conan-libzip package while waiting for official release: https://github.com/chulup/conan-libzip

You'll have to build it yourself, though. Usually simple `python build.py` would be enough but I had to use more complex command:

```
PS C:\workspace\conan-libzip> $env:CONAN_OPTIONS="libzip:shared=False,libzip:with_bzip2=False,libzip:with_openssl=False,libzip:enable_windows_crypto=False"; $env:CONAN_VISUAL_VERSIONS="15,16"; $env:CONAN_VISUAL_RUNTIMES="MT,MTd"; $env:CONAN_BUILD_TYPES="Release,Debug"; $env:CONAN_ARCHS="x86,x86_64"; $env:CONAN_USERNAME="chulup"; $env:CONAN_CHANNEL="testing"; python build.py
```

By the way, that feature works great for me. Thank you, @0-wiz-0 !

**Comment by EZ-Phantom** [comment]
@chulup thx, but I need a little bit more official release due to commercial development. Also I am using cpp :)

**Comment by chulup** [comment]
@EZ-Phantom 
I understand the need for official release.

Sure, I use it for C++ project too, through Conan packet manager. You should try it: my life became a lot easier since I started using it for both 3rd-party libraries and to package my own libs before linking them to internal projects.

**Comment by EZ-Phantom** [comment]
@chulup idk is it okay that we talk about it in this thread :D, but we use docker which is also easy-to-use and simple. Write how to build and install lib once in your image and forget about it until you need to update verison or smth else

**Comment by 0-wiz-0** [comment]
We've just released libzip 1.6.0. Enjoy!

**Comment by EZ-Phantom** [comment]
Thanks!)

пт, 24 янв. 2020 г., 16:05 Thomas Klausner <notifications@github.com>:
